lib/src/front_end/piece_writer.dart

```diff
@@ -242,7 +243,8 @@ class PieceWriter {
   /// Finishes writing and returns a [SourceCode] containing the final output
   /// and updated selection, if any.
   SourceCode finish(Piece rootPiece) {
-    var formatter = Solver(_formatter.pageWidth);
+    var cache = SolutionCache();
```

Comment:
Do we need to initialize SolutionCache outside of the Solver? Can the Solver initialize and just own it's own cache when created?

Reply:
Yeah, doing so lets us reuse cached subtree Solutions across different Solvers. Imagine we have a tree like:

```
Sequence
  Infix
    List
      Element
        ...
  ...
```

1. We create a Solver for the Sequence piece and start trying Solutions.
2. When we get to Infix, it decides to format its operand (List) using a separate Solver.
3. That inner Solver starts trying solutions for List.
4. When it gets to List, that piece decides to format Element using a separate Solver.
5. We find the optimal Solution for that Element subtree and stuff the result in the cache (1).
6. We pop back up to the Solver for Infix. It eventually finds a Solution which gets cached too.
7. Then we pop back to the outermost Solver for the Sequence. It keeps iterating through solutions and eventually decides to format the Infix another way.
8. We go back into Infix. Maybe Infix decides to create another Solver for its operand (List) or maybe it doesn't. Either way, it eventually gets back to Element.
9. Now, when we look up a previously solved format for Element, we find it in the cache, even though the Solver that created it (from step 3) is long gone.

It seems a little convoluted, but I did some testing and I do see that the cache is reusing Solutions created by different Solvers, so it's adding value.